### ISSUE TITLE
Add support for parsing URL query parameters with no value

### ISSUE THREAD

**Issue by calebhailey** [qutheory/vapor#3053]
**Is your feature request related to a problem? Please describe.**

I'm developing an HTTP API using Vapor where some endpoints may offer support for URL query parameters that have a key but no value (e.g. `?foo` vs `?foo=bar`). As far as I can tell, the [URI Spec](https://www.rfc-editor.org/rfc/rfc3986#section-3.4) doesn't specify that query parameters must have a value, so my assumption was that this should be possible. 

**Describe the solution you'd like**

It should be possible to decode query parameters and capture keys with no values:

```swift
var params: Dictionary<String, String?> = try req.query.decode(Dictionary<String, String?>.self)
```

**Describe alternatives you've considered**

This works as expected but it appears to skip query params with no values: 

```swift
var params: Dictionary<String, String> = try req.query.decode(Dictionary<String, String>.self)
```

I've tried switching the Dictionary value type to an optional string and it still won't decode url query parameters without values

```swift
var params: Dictionary<String, String?> = try req.query.decode(Dictionary<String, String?>.self)
```

NOTE: I suspect a Struct attribute with an optional value _would_ work, but I have _not_ tried decoding to a struct because in my case I need support for dynamic query parameter keys, thus the `Dictionary<String, String>`. 

**Additional context**

Vapor appears to parse url params as long as there is an `=`; so `?foo=` works, but `?foo` does not get parsed.

PS – I'm new to Vapor but I'm loving it so far! Keep up the great work! 💪  

EDIT: added link to the URI Spec

**Comment by jhoughjr** [comment]
I've checked URLEncodedFormParser.parse and its behavior appears correct. I havent found the next responsible code in the chain. My suspicion is it makes no sense for the decoder to decode a key as a value maybe. Maybe bogging the key into the value would behave properly by falling through the existing logic. I cant say until I find the chain of execution.
It may even be it doesnt try to decode it as a dictionary of strings as a dictionary with a key and no value is undefined? I'll keep looking.

**Comment by jhoughjr** [comment]
let decoder = _Decoder(data: parsedData, codingPath: [], configuration: configuration)
        return try D(from: decoder) this init and decode appears to be what fails.

**Comment by mkll** [comment]
Are you guys sure that "_the URI Spec doesn't specify that query parameters must have a value_" also mean that it may not have the `=` sign after the key? In my experience, I often see `key=` in the wild, but so far I have never seen just `key`.

**Comment by jhoughjr** [comment]
Checking the RFC it only indicates that key value pairs are commonly used and does not appear to state any requirements on the structure of the query string other than its termination and how to specify pairs. I think we need to confirm this shold be supported before doing much more work. I think it is a conceptual bug more than an implementation bug perhaps.

In my opinion treating flags as bool arguments should be a workaround for getting flags from query params.

**Comment by calebhailey** [comment]
This won't make or break my project, it was more of a curiosity whether this was an oversight/omission in the query parsing code, or an active design decision. I don't have any precedent to refer to in terms of patterns in other projects, so maybe all that suggests this is a bad idea 😅

**Comment by calebhailey** [comment]
But the spec is pretty clear about the query component definition: 

> The query component is indicated by the first question mark ("?") character and terminated by a number sign ("#") character or by the end of the URI.

For reference, I typically defer to MDN docs and/or Javascript implementations as a battle-tested standard, and it appears that JavasScript will parse valueless query parameters. 

```
var u = new URL("https://example.com/?foo=bar&baz")
console.log(u.searchParams) // outputs: URLSearchParams { 'foo' => 'bar', 'baz' => '' }
```

I get the same result in Safari and NodeJS repl.

**Comment by calebhailey** [comment]
Found a workaround via the `Vapor.Request.url` property. I can get at the raw query that way. 

```swift
let u: URLComponents? = URLComponents(string: String(request.url))
```

The `Foundation.URLComponents` class captures query params without values, so it works exactly as I was hoping.

**Comment by jhoughjr** [comment]
I'm not sure if the current behavior is correct or not. I'm not sure what would change.
Since you have a workaround I vote it can be closed unless more specific action is determined.

**Comment by techouse** [comment]
Vapor currently drops key-only query params (e.g. `?flag`) unless there’s an equals sign. If you need this today, [QsSwift](https://github.com/techouse/qs-swift) (my Swift port of Node’s [qs](https://www.npmjs.com/package/qs)) already treats key-only params as present and decodes them sensibly.

From the [test suite](https://github.com/techouse/qs-swift/blob/7e6043671cdb0ec2f59ca90c1a7e93b8e345cbb1/Tests/QsSwiftTests/ExampleTests.swift#L521):
```swift
// ExampleTests.swift — nulls: decoding treats 'a&b=' as empty strings
let r = try Qs.decode("a&b=")
// By default, a key with no "=" is treated like an empty string
#expect((r["a"] as? String) == "")
#expect((r["b"] as? String) == "")
```

If you prefer to [distinguish “no value” from “empty string”, enable strict nulls](https://github.com/techouse/qs-swift/blob/7e6043671cdb0ec2f59ca90c1a7e93b8e345cbb1/Tests/QsSwiftTests/ExampleTests.swift#L534):
```swift
// ExampleTests.swift — nulls: strictNullHandling on decode
let r = try Qs.decode("a&b=", options: .init(strictNullHandling: true))
#expect(r["a"] is NSNull)        // key-only → null
#expect((r["b"] as? String) == "") // "b=" → empty string
```

### Workaround today (in a route/middleware):
```swift
import QsSwift

let raw = req.url.query ?? ""
let params = try Qs.decode(raw)   // [String: Any] with "flag" present
// use `params["flag"]` as "" (or NSNull with strictNullHandling)
```

Longer term, QsSwift could be a pluggable decoder for Vapor’s query parsing so cases like `?flag` are supported natively (either as empty string or null, depending on a strategy option).

**Comment by 0xTim** [comment]
That's not true - we support empty values and map them to `Bool` - https://github.com/vapor/vapor/blob/main/Tests/VaporTests/QueryTests.swift#L294-L297

**Comment by 0xTim** [comment]
QsSwift would also need to work on Linux for it be to used with most Vapor projects

**Comment by techouse** [comment]
> That's not true - we support empty values and map them to Bool - https://github.com/vapor/vapor/blob/main/Tests/VaporTests/QueryTests.swift#L294-L297

I stand corrected 🫡 it got fixed in #3151

> QsSwift would also need to work on Linux for it be to used with most Vapor projects

~~Unfortunately, it doesn't run on Linux due to the lack of `CoreFoundation`/Objective‑C runtime facilities in Swift Foundation on Linux, most notably `NSMapTable`, which the encoder uses for weak, identity‑based cycle detection. [I wrote a whole document explaining why.](https://github.com/techouse/qs-swift/blob/main/LinuxSupport.md)~~

~~I plan to add it in the future, if I find a package that provides the same functionality as `NSMapTable` on Linux.~~

I have now added [experimental support for Linux](https://github.com/techouse/qs-swift?tab=readme-ov-file#linux-support).
